### ISSUE TITLE
Encoding circular objects

### ISSUE THREAD

**Issue by hildjj** [hildjj/node-cbor#35]
``` javascript
a = {}
a.a = a
cbor.encode(a) // RangeError: Maximum call stack size exceeded
```

Mark objects as visited with a symbol, and delete the symbol on a second pass?  Throw an error if the symbol exists on encode.  Alternately: replace value with NULL or other symbol.  See what other libraries do.

**Comment by sandhawke** [comment]
Not sure the best approach here, but I spent much of today playing with it.  For cycle detection I used an array of the things currently being pushed.  No idea how that compares in performance to the Symbol approach.    See code, about 15 lines, at  https://github.com/sandhawke/node-cbor-refs/blob/master/lib/encoder.js#L388

Where it gets interesting is what to do if you find a cycle.  I mostly got cycle detection, serialization, and deserialization working.  I used the cbor tags 28 and 29 from http://cbor.schmorp.de/value-sharing.

I started by adding an encoder option pleaseKeep which is an array of objects which should be emitted, when they are emitted, with the tag 28 ("mark value as (potentially) shared"), and then, thereafter, when they are emitted again, a tag 29 ("reference nth marked value") tag is used in their place.   This lets a higher level, that knows about cycles, flag some bits as needing to be made referenceable like this.

Then I added decoders.   One tricky thing is that I want the back-references to be able to carry across multiple decoder runs, so I pass in 'kept' as an option, but made decodeAllSync do reconstruction by default.   Also, it's pretty tricky code because the references can arrive out-of-order.

Anyway, you can see the code, etc, in my fork, if you're interested.  Probably clearest to start with the test cases: https://github.com/sandhawke/node-cbor-refs/blob/master/test/refs.ava.js

**Comment by sandhawke** [comment]
I implemented this two-pass Symbol approach in my value-sharing fork of borc today.     The Symbol stuff is at: https://github.com/sandhawke/borc-refs/blob/master/src/encoder.js#L417 but it's not clear when you'd want to have it turned on.

See [the readme](https://github.com/sandhawke/borc-refs/blob/master/README.md) for more details on what I did.  It doesn't just detect cycles, but actually serializes and deserializes them.

My first attempt was on a fork of node-cbor, but I got a few things wrong, and then decided to use borc instead.   See [node-cbor-refs](https://github.com/sandhawke/node-cbor-refs/blob/master/README.md) if you're curious.

**Comment by sandhawke** [comment]
Did a bit more work.   Benchmarking showed taking two passes was quite expensive (basically, encoding took 2x times as long, even when I skipped the low-level serialization on the first pass), so I'm wondering if it's okay to by default leave objects 'colored' with a symbol property.   I expect it's harmless.   Also, node coloring works for detecting all kinds of sharing, not just cycles.  Sharing might be desirable, even if it's transmitted as copying.    So, I do node coloring to detect sharing, then if there's sharing (and if caller wants) we'll maintain a path to this node to see if there's a cycle.  (That is, people might want to allow sharing, but cycles are never okay.  Node coloring can't tell the difference, as far as I can figure out.)   

Basic code is here: https://github.com/sandhawke/borc-refs/blob/2e7a3f68442b982ae85882ff800d5f840921849a/src/encoder.js#L445-L473 with of course various surrounding bits.   Again, I know this is in borc not cbor, but the basic issues should be about the same.

**Comment by hildjj** [comment]
This approach makes sense to me, particularly since the encode path is all synchronous already, so we don't have to worry about two happening at once.  I wonder what would happen if we used a CBOR tag to encode what amounts to a weak reference to another part of the tree, so that you could reconstitute the loops on decode.  I think that might require two passes in order to put an identifier on the target, however.  I think I'll start with just detecting loops and aborting, which is better than running out of memory.

**Comment by hildjj** [comment]
Fixed in 89a3b931f618ccf42d40200eed9503cc18208c96.
